oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenIntrospectionEndpointFilter.java

```diff
@@ -0,0 +1,450 @@
+			// will commence via AuthenticationEntryPoint
+			filterChain.doFilter(request, response);
+			return;
+		}
+
+		MultiValueMap<String, String> parameters = OAuth2EndpointUtils.getParameters(request);
+		try {
+			// token (REQUIRED)
+			String token = parameters.getFirst(OAuth2ParameterNames2.TOKEN);
+			if (!StringUtils.hasText(token) || parameters.get(OAuth2ParameterNames2.TOKEN).size() != 1) {
+				throwError(OAuth2ErrorCodes.INVALID_REQUEST, OAuth2ParameterNames2.TOKEN);
+			}
+
+			// token_type_hint (OPTIONAL)
+			Optional<TokenType> tokenTypeHint = Optional
+					.ofNullable(parameters.getFirst(OAuth2ParameterNames2.TOKEN_TYPE_HINT)).map(TokenType::new);
+			if (tokenTypeHint.isPresent() && parameters.get(OAuth2ParameterNames2.TOKEN_TYPE_HINT).size() != 1) {
+				throwError(OAuth2ErrorCodes.INVALID_REQUEST, OAuth2ParameterNames2.TOKEN_TYPE_HINT);
+			}
+			TokenIntrospectionSuccessResponse tokenIntrospectionResponse = generateTokenIntrospectionResponse(principal,
+					token, tokenTypeHint);
+
+			this.sendTokenIntrospectionResponse(response, tokenIntrospectionResponse);
+		} catch (OAuth2AuthenticationException ex) {
+			SecurityContextHolder.clearContext();
+			sendErrorResponse(response, ex.getError());
+		}
+	}
+
+	private TokenIntrospectionSuccessResponse generateTokenIntrospectionResponse(Authentication principal, String token,
+			Optional<TokenType> tokenTypeHint) {
+		try {
+			// get client from token
+			OAuth2Authorization authorization = findAuthorizationByToken(token, tokenTypeHint)
+					.orElseThrow(() -> new InvalidTokenException("Token not found"));
+
+			// check if token corresponds to authorized Client
+			OAuth2ClientAuthenticationToken clientAuthentication = principal instanceof OAuth2ClientAuthenticationToken
+					? (OAuth2ClientAuthenticationToken) principal
+					: null;
+			if (clientAuthentication == null || !clientAuthentication.getRegisteredClient().getId()
+					.equals(authorization.getRegisteredClientId())) {
+				throw new InvalidTokenException("Token does not correspond to authenticated client");
+			}
+
+			// we obtained the authorization from the token, no need to validate Optional here
+			AbstractOAuth2Token oauthToken = authorization.getTokens().getToken(token).get();
+			if (authorization.getTokens().getTokenMetadata(oauthToken).isInvalidated()) {
+				throw new InvalidTokenException("Token has been invalidated");
+			}
+
+			oauthToken = parseJwt(oauthToken);
+
+			validateToken(oauthToken);
+
+			TokenIntrospectionSuccessResponse.Builder builder = new TokenIntrospectionSuccessResponse.Builder(true);
+			builder.clientID(new ClientID(clientAuthentication.getRegisteredClient().getClientId()));
+			Optional.ofNullable(oauthToken.getIssuedAt()).map(Date::from).ifPresent(builder::issueTime);
+			Optional.ofNullable(oauthToken.getExpiresAt()).map(Date::from).ifPresent(builder::expirationTime);
+			TokenToIntrospectionResponseFieldsMapper.extractFromToken(oauthToken, builder);
```

Comment:
I have added functionality to retrieve all the response fields, except for [the _username_](https://tools.ietf.org/html/rfc7662#section-2.2) ( _Human-readable identifier for the resource owner who authorized this token._).
For when we have a _Jwt_, we could use the "sub" field, but for the rest of the tokens, I think we're not storing that information anywhere.
Let me know what you think